src/Agent.Worker/TaskRestrictionsChecker.cs

```diff
@@ -0,0 +1,143 @@
+﻿using Agent.Sdk.Knob;
+using Microsoft.TeamFoundation.Common;
+using Microsoft.TeamFoundation.DistributedTask.WebApi;
+using Microsoft.VisualStudio.Services.Agent.Util;
+using Microsoft.VisualStudio.Services.Agent.Worker.Telemetry;
+using Microsoft.VisualStudio.Services.WebPlatform;
+using Newtonsoft.Json;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Microsoft.VisualStudio.Services.Agent.Worker
+{
+    [ServiceLocator(Default = typeof(TaskRestrictionsChecker))]
+    public interface ITaskRestrictionsChecker : IAgentService
+    {
+        bool CheckCommand(IExecutionContext context, IWorkerCommand workerCommand, Command command);
+        bool CheckSettableVariable(IExecutionContext context, string variable);
+    }
+
+    public sealed class TaskRestrictionsChecker : AgentService, ITaskRestrictionsChecker
+    {
+        public bool CheckCommand(IExecutionContext context, IWorkerCommand workerCommand, Command command)
+        {
+            ArgUtil.NotNull(context, nameof(context));
+            ArgUtil.NotNull(workerCommand, nameof(workerCommand));
+            ArgUtil.NotNull(command, nameof(command));
+
+            return Check(
+                context,
+                restrictions => !restrictions.IsCommandAllowed(workerCommand),
+                () => context.Warning(StringUtil.Loc("CommandNotAllowed", command.Area, command.Event)),
+                restrictions => PublishCommandTelemetry(context, restrictions, command));
+        }
+
+        public bool CheckSettableVariable(IExecutionContext context, string variable)
+        {
+            ArgUtil.NotNull(context, nameof(context));
+            ArgUtil.NotNull(variable, nameof(variable));
+
+            return Check(
+                context,
+                restrictions => !restrictions.IsSetVariableAllowed(variable),
+                () => context.Warning(StringUtil.Loc("SetVariableNotAllowed", variable)),
+                restrictions => PublishVariableTelemetry(context, restrictions, variable));
+        }
+
+        private bool Check(
+            IExecutionContext context,
+            Func<TaskRestrictions, bool> predicate,
+            Action warn,
+            Action<TaskDefinitionRestrictions> publishTelemetry)
+        {
+            ArgUtil.NotNull(context, nameof(context));
+            ArgUtil.NotNull(predicate, nameof(predicate));
+            ArgUtil.NotNull(warn, nameof(warn));
+            ArgUtil.NotNull(publishTelemetry, nameof(publishTelemetry));
+
+            var matches = context.Restrictions?.Where(predicate);
```

Comment:
It took me a bit to grok what was happening here. I think this file would be more readable if this line was pulled into the callers, and instead of passing in `Func<TaskRestrictions, bool> predicate` we passed in `TaskRestrictions matches` (or I'd probably actually call it `restrictions` at that point. I'd push for that change, but if you don't do that I'd recommend 2 other changes:
1) Give `predicate` a  more clear name
2) Above change `restrictions => !restrictions.IsSetVariableAllowed(variable),` to `restriction => !restriction.IsSetVariableAllowed(variable)` - restrictions shouldn't be plural because in the context where that function is called we're operating on a single restriction

Reply:
`predicate` is a pretty commonly used word in functional programming and the word the LINQ docs use for the Where method.

The class is called TaskRestrictions (plural) because each instance can contain multiple restrictions (command, settable variables). So `restrictions` as an instance of `TaskRestrictions` is correct.

Reply:
> predicate is a pretty commonly used word in functional programming and the word the LINQ docs use for the Where method.

That's true, but in this case it doesn't really give us any information about what the predicate is used for. To me, its kinda like passing an int and naming it number. I'd prefer something that actually connotes the purpose of this filter. 

I think my point here is that reading this method by itself doesn't give you enough context to know what this statement is actually doing, when in reality its always doing the same thing (filtering out TaskRestrictions that are applied but don't result in the command actually needing to be restricted). Likewise, we also don't have enough information to call this function without knowing something about the internals (namely that the predicate should filter the tasks down to only the relevant restrictions).

I don't think this is wrong, but I think it would be clearer and more in line with the rest of the codebase to push it into the original callers

> The class is called TaskRestrictions (plural) because each instance can contain multiple restrictions (command, settable variables). So restrictions as an instance of TaskRestrictions is correct.

That's fair

Reply:
I added type declarations to the lambas to make the type of the `restrictions` args clearer.

I didn't want to duplicate the filtering of the matches, but it is a fair point that the fact that the predicates were for filtering wasn't clear. I inverted the logic so the predicate is now the check, which I think is clearer.